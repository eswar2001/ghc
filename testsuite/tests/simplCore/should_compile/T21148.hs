module T211148 where

-- The point of this test is that f should get a (nested)
-- CPR property, with a worker of type
-- $wf :: Int# -> State# RealWorld -> (# State# RealWorld, Int# #)

{-# NOINLINE f #-}
-- The NOINLINE makes GHC do a worker/wrapper split
-- even though f is small
f :: Int -> IO Int
f x = return $! sum [0..x]

