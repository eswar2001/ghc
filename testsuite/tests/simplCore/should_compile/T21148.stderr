
==================== Tidy Core ====================
Result size of Tidy Core
  = {terms: 73, types: 80, coercions: 6, joins: 2/2}

-- RHS size: {terms: 1, types: 0, coercions: 0, joins: 0/0}
T211148.$trModule4 :: GHC.Prim.Addr#
[GblId,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True, Guidance=IF_ARGS [] 20 0}]
T211148.$trModule4 = "main"#

-- RHS size: {terms: 2, types: 0, coercions: 0, joins: 0/0}
T211148.$trModule3 :: GHC.Types.TrName
[GblId,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True, Guidance=IF_ARGS [] 10 10}]
T211148.$trModule3 = GHC.Types.TrNameS T211148.$trModule4

-- RHS size: {terms: 1, types: 0, coercions: 0, joins: 0/0}
T211148.$trModule2 :: GHC.Prim.Addr#
[GblId,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True, Guidance=IF_ARGS [] 30 0}]
T211148.$trModule2 = "T211148"#

-- RHS size: {terms: 2, types: 0, coercions: 0, joins: 0/0}
T211148.$trModule1 :: GHC.Types.TrName
[GblId,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True, Guidance=IF_ARGS [] 10 10}]
T211148.$trModule1 = GHC.Types.TrNameS T211148.$trModule2

-- RHS size: {terms: 3, types: 0, coercions: 0, joins: 0/0}
T211148.$trModule :: GHC.Types.Module
[GblId,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True, Guidance=IF_ARGS [] 10 10}]
T211148.$trModule
  = GHC.Types.Module T211148.$trModule3 T211148.$trModule1

-- RHS size: {terms: 41, types: 35, coercions: 0, joins: 2/2}
T211148.$wf [InlPrag=NOINLINE]
  :: GHC.Prim.Int#
     -> GHC.Prim.State# GHC.Prim.RealWorld
     -> (# GHC.Prim.State# GHC.Prim.RealWorld, GHC.Prim.Int# #)
[GblId, Arity=2, Str=<L><L>, Unf=OtherCon []]
T211148.$wf
  = \ (ww_s179 :: GHC.Prim.Int#)
      (eta_s17b [OS=OneShot] :: GHC.Prim.State# GHC.Prim.RealWorld) ->
      case GHC.Prim.># 0# ww_s179 of {
        __DEFAULT ->
          join {
            exit_X0 [Dmd=SC(S,C(1,!P(L,L)))]
              :: GHC.Prim.Int#
                 -> GHC.Prim.Int#
                 -> (# GHC.Prim.State# GHC.Prim.RealWorld, GHC.Prim.Int# #)
            [LclId[JoinId(2)(Nothing)], Arity=2, Str=<L><L>]
            exit_X0 (x_s16Z [OS=OneShot] :: GHC.Prim.Int#)
                    (ww1_s172 [OS=OneShot] :: GHC.Prim.Int#)
              = (# eta_s17b, GHC.Prim.+# ww1_s172 x_s16Z #) } in
          joinrec {
            $wgo3_s175 [InlPrag=[2], Occ=LoopBreaker, Dmd=SC(S,C(1,!P(L,L)))]
              :: GHC.Prim.Int#
                 -> GHC.Prim.Int#
                 -> (# GHC.Prim.State# GHC.Prim.RealWorld, GHC.Prim.Int# #)
            [LclId[JoinId(2)(Nothing)], Arity=2, Str=<L><L>, Unf=OtherCon []]
            $wgo3_s175 (x_s16Z :: GHC.Prim.Int#) (ww1_s172 :: GHC.Prim.Int#)
              = case GHC.Prim.==# x_s16Z ww_s179 of {
                  __DEFAULT ->
                    jump $wgo3_s175
                      (GHC.Prim.+# x_s16Z 1#) (GHC.Prim.+# ww1_s172 x_s16Z);
                  1# -> jump exit_X0 x_s16Z ww1_s172
                }; } in
          jump $wgo3_s175 0# 0#;
        1# -> (# eta_s17b, 0# #)
      }

-- RHS size: {terms: 14, types: 19, coercions: 0, joins: 0/0}
T211148.f1 [InlPrag=NOINLINE[final]]
  :: Int
     -> GHC.Prim.State# GHC.Prim.RealWorld
     -> (# GHC.Prim.State# GHC.Prim.RealWorld, Int #)
[GblId,
 Arity=2,
 Str=<1!P(L)><L>,
 Cpr=1(, 1),
 Unf=Unf{Src=StableSystem, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=True,boring_ok=False)
         Tmpl= \ (x_s177 [Occ=Once1!] :: Int)
                 (eta_s17b [Occ=Once1, OS=OneShot]
                    :: GHC.Prim.State# GHC.Prim.RealWorld) ->
                 case x_s177 of { GHC.Types.I# ww_s179 [Occ=Once1] ->
                 case T211148.$wf ww_s179 eta_s17b of
                 { (# ww1_s17e [Occ=Once1], ww2_s17j [Occ=Once1] #) ->
                 (# ww1_s17e, GHC.Types.I# ww2_s17j #)
                 }
                 }}]
T211148.f1
  = \ (x_s177 :: Int)
      (eta_s17b [OS=OneShot] :: GHC.Prim.State# GHC.Prim.RealWorld) ->
      case x_s177 of { GHC.Types.I# ww_s179 ->
      case T211148.$wf ww_s179 eta_s17b of { (# ww1_s17e, ww2_s17j #) ->
      (# ww1_s17e, GHC.Types.I# ww2_s17j #)
      }
      }

-- RHS size: {terms: 1, types: 0, coercions: 6, joins: 0/0}
f [InlPrag=NOINLINE[final]] :: Int -> IO Int
[GblId,
 Arity=2,
 Str=<1!P(L)><L>,
 Cpr=1(, 1),
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=True)}]
f = T211148.f1
    `cast` (<Int>_R %<'Many>_N ->_R Sym (GHC.Types.N:IO[0] <Int>_R)
            :: (Int
                -> GHC.Prim.State# GHC.Prim.RealWorld
                -> (# GHC.Prim.State# GHC.Prim.RealWorld, Int #))
               ~R# (Int -> IO Int))



